### PR TITLE
feat(curriculum); added another challenge to help teach recursion

### PR DESCRIPTION
**Pull request by rns350**
- [x ] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x ] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x ] My pull request targets the `master` branch of freeCodeCamp.
- [x ] None of my changes are plagiarized from another source without proper attribution.
- [x ] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x ] My changes do not use shortened URLs or affiliate links.

Addresses #16107

### CODE REVIEW

**Comment by RandellDawson** [comment]
@rns350 I think this is a good start, but see my feedback below.

1. You really need to clean up the git history.  It appears when you first started working on the branch, your master branch (which you should have branched off of) was not clean (did not have the latest updates).  This is probably why you had different package.json and package-lock.json files to start with.  I suggest rebasing against `master` and which should reduce the number of commits showing.  Also, I recommend squashing as many of the commits as makes sense.

2. Not sure the order in which this challenge should appear, but you will need to update [this meta.json file](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/curriculum/challenges/_meta/basic-javascript/meta.json) when you do decide.

3. I would remove the example in the editor and make sure your description section has any examples you feel are important.

4. The name of the challenge needs to follow the guidelines as indicated in the [styling guide](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/docs/style-guide-for-curriculum-challenges.md#naming-challenges)

5. I would rename the function to something more description and use camelCase for the function name.

6. I feel like the example should be more closely tied to the challenge to be completed. There would only need to be slight changes in logic to accomplish the task, yet still achieve the desired goal of teaching the recursion concept.

In an attempt to show you what I mean above, I have created a separate PR (#35968) with a different example and problem for the camper to complete.  I am not saying that my challenge should be considered instead of yours.  Instead, I am hoping you could take some ideas from mine (the similarities of the example and the problem to be solved) along with the kinds of tests I wrote to validate the solution.

**Comment by rns350** [comment]
Alright, thankyou for the advice, and sorry about the formatting errors!  I took a look at what you submitted and see what you mean.  I'll make some changes to my file and resubmit a PR before the end of the week.  Thanks for your time, I really appreciate it!

**Comment by RandellDawson** [comment]
You don't have to create a new PR.  Just update your changes on this one.  Also, make sure to rebase against master so you clean up the history.

**Comment by RandellDawson** [comment]
@rns350 Use `git pull --rebase upstream master` which should clean up your history a bit.  If that does not bring the number of commits down, then let me know.

**Comment by rns350** [comment]
Hey Randell, thanks for the advice.  Just finished editing the challenge and am working on cleaning up the commit history right now.  Will get back to you shortly about the rebase status

**Comment by RandellDawson** [comment]
@rns350 It appears that the master branch of your fork has been spoiled.  It probably is because you made commits at some point to the master branch instead of a separate branch.  @raisedadead I attempted to rebase this PR, but it had over 50 conflicts in the `package-lock.json` file that it did not seem worth trying to resolve.  What is the easiest way to clean up this git history for this PR.  It has commits unrelated to the two files.  I understand they have been reverted, but it is a mess in my opinion.

**Comment by rns350** [comment]
Hi Randell.  I just managed to do the rebase and skip those commits, so they are gone in the history of the branch now.  The only commits that are shown now involve the creation and updating of the challenge, and the updating of the meta.json file

**Comment by RandellDawson** [comment]
You still have commits that changed the package.json and package-lock.json and then reverted the changed too.

**Comment by RandellDawson** [comment]
Go ahead squash your commits to one.

**Comment by rns350** [comment]
All of the commits are squashed and ready to be pulled!

**Comment by RandellDawson** [comment]
You can get rid of the following since you are not using it.
````
### Before Test
<div id='{ext}-setup'>

```js
```

</div>
````

**Comment by rns350** [comment]
All set!  Resquashed the commits into one

**Comment by raisedadead** [comment]
> [...] What is the easiest way to clean up this git history for this PR. [...]

Can say since the exact approach since this PR looks cleaned up now  :). 

But one approach I can think by assuming the context is: 
- Branch off the master (doesn't matter if soiled) to a new `fix/branch-name` branch. 
- Hard reset this new branch to `upstream master` 
- Then use `cherry-pick` to grab that one commit which was intended to make the change the first place, or manually make a change a make a fresh commit if its a small quick change.
- Delete original the PR branch locally `git branch -D pr-branch-name`.
- Create the pr branch again from your 'fix' branch and force push to `origin` to update the PR.
- Go back to the master and hard reset it too to `upstream master` for future work.

**Comment by rns350** [comment]
wait so do I still need to clean up the commit history?  As far as I can tell, the squashed commit only shows two modified files now, the new recursion challenge and the meta file.

**Comment by RandellDawson** [comment]
@rns350 No you are fine now.  We just need to review the PR now.

**Comment by rns350** [comment]
Awesome thanks so much guys!

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/introducing-recursion-as-a-solution-to-iterative-problems.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/introducing-recursion-as-a-solution-to-iterative-problems.english.md]:
```suggestion
  padArrayBackwards(arr, n);
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/introducing-recursion-as-a-solution-to-iterative-problems.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/introducing-recursion-as-a-solution-to-iterative-problems.english.md]:
```suggestion
function padArrayBackwards(myArray, n){
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/introducing-recursion-as-a-solution-to-iterative-problems.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/introducing-recursion-as-a-solution-to-iterative-problems.english.md]:
```suggestion
    padArrayBackwards(myArray, n - 1);
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Thanks for the PR!  While there are lot of suggestions, they're almost all replacing 'fill' with 'pad', since `fill` [already exists ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill) in JS and means 'overwrite stuff in an array' not 'append stuff onto an array'.

**Review by thecodingaviator — COMMENTED** [review]
Hi,

As there has been no response from your end for a while, we are closing this PR.

Feel free to request this PR to be reopened, if you still wish to make the changes you originally added.  Also, remember to update your fork, so you have the latest version of master in case it has changed significantly.

Happy Coding! ✈️

**Comment by RandellDawson** [comment]
@ojeytonwilliams I made most of the changes you suggested and since we already have the recursive challenge I created with all the "intro" parts, I shorted the description of this one and also renamed this challenge to be called `Use Recursion to Create a Countdown".  Feel free to give critiques on these changes.

Honestly, I still would prefer to have a function which did not require the two parameters (like the version below):
```
function countdown(n){
  if(n === 1){
    return [1];
  }
  else{
    return [n].concat(countdown( n - 1));
  }
}

countdown(5); // [5, 4, 3, 2, 1]
```
Another approach for this 3rd recursion challenge would be to introduce the idea of a helper function, so that the main function only has the single parameter `n` and then the helper function does the actual recursive work.  However, at this point, I think we should just try to get this one merged and see what kind of feedback we get on the forum and Gitter Help room once it gets deployed to production.

File: curriculum/challenges/_meta/basic-javascript/meta.json
Comment (ojeytonwilliams) [inline curriculum/challenges/_meta/basic-javascript/meta.json]:
```suggestion
      "Use Recursion to Create a Countdown"        
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md]:
```suggestion
We have defined a function called <code>countdown</code> with two parameters.  The function should take an array in the <code>myArray</code> parameter and append the numbers n through 1 based on the <code>n</code> parameter.  
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
I found a couple of things, but it looks good otherwise.

File: curriculum/challenges/_meta/basic-javascript/meta.json
Comment (RandellDawson) [reply curriculum/challenges/_meta/basic-javascript/meta.json]: Created a new commit for this change.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md
Comment (RandellDawson) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/use-recursion-to-create-a-countdown.md]: Created a new commit for this change.

**Review by thecodingaviator — DISMISSED** [review]
@rns350 I think you should add the guide file for this challenge in this same PR. 


Feel free to dismiss my review if any other mod has a different opinion

**Comment by RandellDawson** [comment]
@thecodingaviator The guide is no longer being built. All hint/solutions for challenges have to be created on the forum.  Until we have a more formal process documented, I am manually creating the forum articles and then adding the forumTopicId to the challenge frontmatter once each new challenge is merged.  We only have a few more new challenges coming from PRs, so it will not be a big deal to handle these as they are merged.  We are no longer accepting any certification folder PRs.  Also, we are no longer accepting any new PRs which add new challenges to the current curriculum.  All new challenges would be added through the new curriculum repo.

@ojeytonwilliams Is this PR good to go now?

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by RandellDawson** [comment]
@scissorsneedfoodtoo Just need your approval and then you can merge it.

**Comment by scissorsneedfoodtoo** [comment]
These recursion challenges look great! Thank you for adding this challenge, @rns350, and for your review @RandellDawson, @ojeytonwilliams, and @thecodingaviator. Looking forward to seeing this live soon.

**Review by scissorsneedfoodtoo — APPROVED** [review]

**Pull request merged**
